test/controllers/ContractorDetailsControllerSpec.scala

```diff
@@ -0,0 +1,127 @@
+import play.api.http.Status.{BAD_REQUEST, SEE_OTHER}
+import play.api.test.Helpers.status
+import support.ControllerUnitTest
+import support.builders.models.UserBuilder.aUser
+import support.builders.models.mongo.CisUserDataBuilder.aCisUserData
+import support.mocks.{MockAuthorisedAction, MockContractorDetailsService, MockErrorHandler}
+import utils.InYearUtil
+import views.html.ContractorDetailsView
+
+import scala.concurrent.Future
+
+class ContractorDetailsControllerSpec extends ControllerUnitTest
+  with MockAuthorisedAction
+  with MockContractorDetailsService
+  with MockErrorHandler {
+
+  private val pageView = inject[ContractorDetailsView]
+
+  val inYearAction = new InYearUtil
+
+  private val underTest = new ContractorDetailsController()(cc, mockAuthorisedAction, pageView,
+    mockContractorDetailsService, inYearAction, ec, appConfig)
+
+  ".show" should {
+    "redirect to UnauthorisedUserErrorController when authentication fails" in {
+      mockFailToAuthenticate()
+
+      await(underTest.show(taxYear = taxYearEOY, None)(fakeIndividualRequest)) shouldBe
+        Redirect(UnauthorisedUserErrorController.show())
+    }
+
+    "redirect when in year" in {
+      mockAuthAsIndividual(Some(aUser.nino))
+      await(underTest.show(taxYear = taxYear, None)(fakeIndividualRequest)) shouldBe
+        Redirect(appConfig.incomeTaxSubmissionOverviewUrl(taxYear))
+    }
+
+    "redirect when service returns left" in {
+      mockAuthAsIndividual(Some(aUser.nino))
+      mockCheckAccessContractorDetailsPage(taxYearEOY, aUser, Future(Left("String")))
+      await(underTest.show(taxYear = taxYearEOY, None)(fakeIndividualRequest)) shouldBe
+        Ok("String")
+    }
+
+    "Show view when service returns Right(None)" in {
+      mockAuthAsIndividual(Some(aUser.nino))
+      mockCheckAccessContractorDetailsPage(taxYearEOY, aUser, Future(Right(None)))
+      await(underTest.show(taxYear = taxYearEOY, None).apply(fakeIndividualRequest)).header.status shouldBe Ok.header.status
+    }
+
+    "Show view when service returns Right(Some)" in {
+      mockAuthAsIndividual(Some(aUser.nino))
+      mockCheckAccessContractorDetailsPage(taxYearEOY, aUser, Future(Right(Some(aCisUserData))))
+      await(underTest.show(taxYear = taxYearEOY, None).apply(fakeIndividualRequest)).header.status shouldBe Ok.header.status
+    }
+  }
+
+  ".submit" should {
+    "redirect to UnauthorisedUserErrorController when authentication fails" in {
+      mockFailToAuthenticate()
+
+      await(underTest.submit(taxYear = taxYearEOY)(fakeIndividualRequest)) shouldBe
+        Redirect(UnauthorisedUserErrorController.show())
+    }
+
+    "redirect when in year" in {
+      mockAuthAsIndividual(Some(aUser.nino))
+      await(underTest.submit(taxYear = taxYear)(fakeIndividualRequest)) shouldBe
+        Redirect(appConfig.incomeTaxSubmissionOverviewUrl(taxYear))
+    }
+
+    "return bad request when form is broken" in {
+      val error = new DatabaseError {
```

Comment:
you can use one of the database errors instead of a string e.g. DataNotUpdatedError